### ISSUE TITLE
No sendfile on OS X / MacOS

### ISSUE THREAD

**Issue by tailhook** [nix-rust/nix#560]
I guess this is just not implemented yet, right? Not something that's impossible or inviable to implement.

The problem might be that the system call is very different on OS X and linux. So should it be implemented on each system as is? Or should some common denominator be in nix? (given that syscall exists in libc for both platforms with different signature)

**Comment by Mic92** [comment]
Here is how go handle this: https://golang.org/search?q=sendfile

**Comment by kamalmarhubi** [comment]
Sorry for the delayed response. I think the right way to handle this in nix is to have each platforms syscall implemented as-is, wrapping the libc implementations. Happy to accept a PR!

**Comment by morrowa** [comment]
This is somewhat confusing as there is an identically named method in BSDs (including [FreeBSD](https://www.freebsd.org/cgi/man.cgi?query=sendfile&sektion=2) and [Darwin](https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man2/sendfile.2.html)) which has different semantics.

The BSD `sendfile` copies the contents of a file descriptor to a streaming socket whereas the Linux `sendfile` copies from one file descriptor to another.

What's the convention for handling this scenario? Should the BSD `sendfile` also be exposed under a different name?

Once this question is answered, I'd be interested in taking this ticket.

**Comment by asomers** [comment]
They're not as different as they sound.  A streaming socket, after all, _is_ a file descriptor.  Still, there are important differences, like the `sf_hdtr` structure that is absent on Linux or the `sbytes` argument that is absent on OSX.  I agree with @kamalmarhubi that it would probably be best for Nix to wrap `sendfile` in its native form for each operating system.  Combining them all into a single cross-platform API is beyond the remit of Nix.

**Comment by morrowa** [comment]
OK! Would wrapping the Darwin syscall implementation have any value? E.g. [what Go does in their syscall package](https://golang.org/src/syscall/syscall_darwin_amd64.go?h=sendfile#L53)?

**Comment by asomers** [comment]
Oh, I don't want to get into the business of calling bare syscalls.  That's caused a lot of trouble for Go.  Better to use the C APIs.

**Comment by kamalmarhubi** [comment]
I agree on the above. We don't have the resources to deal with the potential issues that come from bare syscalls.

I think the only time we should consider it is where the libc doesn't expose a wrapper. The cases I know of in nix are the linux calls `memfd_create` and `pivot_root`, for which glibc has no wrapper. (Well, apparently it does as of 2.27, but at two months old that's *way* too new.)

**Comment by morrowa** [comment]
Sounds good. Could we modify this ticket to include both Darwin and the other BSDs, since it's the same API? I'll get started on this tonight.

**Comment by morrowa** [comment]
This merged years ago in #901 and can be closed as complete.
